### PR TITLE
GH-14844: [Java] Short circuit null checks when comparing non null field types

### LINKED ISSUE #14844: [Java] VectorValueComparator should skip the null test for NonNullable FieldTypes types
### Describe the enhancement requested

I recently ran a profiler on a simple implementation of a sort and noticed that the VectorValueComparator spends quite a lot of time checking the values to be compared for `null` values. In this case I had already declared the `FieldType` to be `notNullable`. Unless I am misunderstanding something, the values cannot be `null` so the comparator is making unnecessary and expensive comparisons.

As a test I made a comparator that skips the null tests and runs ~12% faster as a result. Based on this I think the `VectorValueComparator` should skip the `nul`l test for NonNullable `FieldTypes`.

I'd be happy to contribute a change to do this if you think it would be reasonable.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by markjschreiber**
Relates to #14844 

To avoid expensive checks for null values when the vectors cannot contain nulls, a flag is set when a vector is attached to indicate if a null check is needed when `compare(idx1, idx2)` is called. If it isn't then the call will immediately redirect to `compareNotNull(index1, index2)`.
* Closes: #14844

### CODE REVIEW

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java
Comment (markjschreiber) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java:93]: For the cases I have tested the java vectors seem to have a validity buffer even when there can be no nulls. Depending on the `AllocationManager` this is some form of buffer called `EMPTY` which is not null but cannot allocate memory for items. It's not possible the test if the buffer is equal to one of these because they are `private`

However, given that someone might change the implementation one day I will include a check for a `null` validity buffer to catch this case as well.

Comment (lidavidm) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java]: Ah, that's what I wasn't remembering. In that case, the right check might be to see if the buffer length is 0 (in which case there can't be any nulls)?

Comment (markjschreiber) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java]: Weirdly it seems that the validity buffer does have a capacity set when any values are added to the value buffer. There is a method to find the number of nulls in the vector which is relatively fast, only needs to be checked when attaching the vector and can be skipped for `FieldTypes` that are `nonNullable`

```
private boolean mayHaveNulls(V v) {
    return  v.getField().getFieldType().isNullable() ||
            v.getNullCount() > 0;
  }
```

**Comment by cyb70289** [comment]
@markjschreiber ping me if you need CI approval. Most maintainers are in vacation.

**Review by lidavidm — APPROVED** [review]
Thanks!

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java
Comment (lidavidm) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java]: Ah, thanks for working through it. I suppose Java doesn't optimize the validity buffer in the same way that C++ does.

**Pull request merged**
